crates/biome_deserialize/src/lib.rs

```diff
@@ -1,31 +1,346 @@
+    fn range(&self) -> TextRange;
+
+    /// Returns the deserialized form of this value using `visitor`.
+    /// Any diagnostics emitted during deserialization are appended to `diagnostics`.
+    fn deserialize<V: DeserializationVisitor>(
+        self,
+        visitor: V,
+        diagnostics: &mut Vec<DeserializationDiagnostic>,
+    ) -> Option<V::Output>;
+}
+
+/// This trait represents a visitor that walks through a [DeserializableValue].
+///
+/// We assume that a deserializable value has one of the following type:
+/// - null / none
+/// - boolean
+/// - number (integer, floats)
+/// - string
+/// - array
+/// - map (key-value pairs with value's type that can depend on its key)
+///
+/// Every type is associated to a `visit_` method.
+/// [DeserializableValue::deserialize] calls the `viist_` method that matches the type of the value.
+///
+/// Most of the time you should implement [Deserializable] and rely on existing deserializable types.
+/// You should use a visitor only when you deserialize a map or a union of several types.
+///
+/// ### Examples
+///
+/// ```
+/// use biome_deserialize::{DeserializationDiagnostic,  Deserializable, DeserializableValue, DeserializationVisitor, ExpectedType};
+/// use biome_deserialize::json::deserialize_from_json_str;
+/// use biome_rowan::{TextRange, TokenText};
+///
+/// struct Person {
+///     name: String
+/// }
+///
+/// impl Deserializable for Person {
+///     fn deserialize(
+///         value: impl DeserializableValue,
+///         diagnostics: &mut Vec<DeserializationDiagnostic>,
+///     ) -> Option<Self> {
+///         value.deserialize(PersonVisitor, diagnostics)
+///     }
+/// }
+///
+/// struct PersonVisitor;
+/// impl DeserializationVisitor for PersonVisitor {
+///     type Output = Person;
+///     const EXPECTED_TYPE: ExpectedType = ExpectedType::MAP;
+///
+///     fn visit_map(
+///         self,
+///         members: impl Iterator<Item = (impl DeserializableValue, impl DeserializableValue)>,
+///         range: TextRange,
+///         diagnostics: &mut Vec<DeserializationDiagnostic>,
+///     ) -> Option<Self::Output> {
+///         const ALLOWED_KEYS: &[&str] = &["name"];
+///         let mut name = None;
+///         for (key, value) in members {
+///             let key_range = key.range();
+///             let Some(key) = TokenText::deserialize(key, diagnostics) else {
+///                 continue;
+///             };
+///             match key.text() {
+///                 "name" => {
+///                     name = Deserializable::deserialize(value, diagnostics);
+///                 },
+///                 _ => diagnostics.push(DeserializationDiagnostic::new_unknown_key(
+///                     key.text(),
+///                     key_range,
+///                     ALLOWED_KEYS,
+///                 )),
+///             }
+///         }
+///         Some(Person { name: name? })
+///     }
+/// }
+///
+/// enum Union {
+///     Bool(bool),
+///     Str(String),
+/// }
```

Comment:
This example is complicated to understand because it doesn't show how's the original data format. I think we should mention the data format, so we can explain why the union use it. Also, we should explain how the visitor works when we have a union, I mean under the hood. There's a lot of magic going on.

Reply:
Do you think we can use JSON? or should I introduce a fake data format?

Reply:
I think JSON is a good example. Maybe we can have more examples in the `json` module, and we can redirect users to that crate to have more "real world" examples.